KEYCLOAK-9572 Support for multiple CRLs with X509 authentication

mposolda opened the pull request:
<!---
Please read https://github.com/keycloak/keycloak/blob/master/CONTRIBUTING.md and follow these guidelines when contributing to Keycloak
-->

mposolda committed:
KEYCLOAK-9572 Support for multiple CRLs with X509 authentication

hmlnarik commented on server-spi-private/src/main/java/org/keycloak/models/Constants.java:
```diff
@@ -73,5 +74,12 @@
 
+    // Delimiter to be used in the configuration of authenticators (and some other components) in case that we need to save
+    // multiple values into single string
+    String CFG_DELIMITER = "##";
```
`public static`

Since `Constants` is an interface, this is not an option, and the Pattern below would be instantiated with every `Constants` instantiations. Could this be placed rather to a proper class or - even better - `Constants` be turned into a final class?

mposolda replied:
I don't think this will be the case as fields declared on interface are automatically considered with modifiers "public static final" . Even if there are some instantiations of objects, which inherit from Constants interface, the pattern will be compiled just once. However AFAIK "constant interface" is not good pattern and I don't like it too, so I treat this as an opportunity to change this to "final class" :)

hmlnarik replied:
You're indeed right, I should stop doing reviews that late in the evening :-) Double thanks for turning this interface into a final class!

hmlnarik commented on server-spi-private/src/main/java/org/keycloak/models/Constants.java:
```diff
@@ -73,5 +74,12 @@
+    String CFG_DELIMITER = "##";
+
+    // Better performance to use this instead of String.split
+    Pattern CFG_DELIMITER_PATTERN = Pattern.compile(CFG_DELIMITER);
```
`public static`

Would be better to ignore whitespace around the delimiter? `"\\s*##\\s*"`?

mposolda replied:
Fixed, Thanks!

hmlnarik commented on services/src/main/java/org/keycloak/authentication/authenticators/x509/AbstractX509ClientCertificateAuthenticatorFactory.java:
```diff
@@ -138,14 +139,18 @@
-        cRLRelativePath.setHelpText("The path to a CRL file that contains a list of revoked certificates. Paths are assumed to be relative to $jboss.server.config.dir");
+        cRLRelativePath.setLabel("CRL Path");
+        cRLRelativePath.setHelpText("Applied just if CRL checking is ON and CRL Distribution point is OFF. It contains the URL (typically 'http' or 'ldap') " +
+                " where the CRL is available. Alternatively it can contain the path to a CRL file that contains a list of revoked certificates. Paths are assumed to be relative to $jboss.server.config.dir ." +
```
Together with end of previous row, two spaces; at the end there should be dot and space swapped.
```suggestion
                "where the CRL is available. Alternatively it can contain the path to a CRL file that contains a list of revoked certificates. Paths are assumed to be relative to $jboss.server.config.dir. " +
```

mposolda replied:
Fixed, Thanks

hmlnarik reviewed (commented):
Thanks, overall looks good, see comments inline.

mposolda commented:
@hmlnarik Thanks, I've fixed per your suggestion. I also removed the X509CRLTest class as it was duplication of the class X509BrowserCRLTest (Before I've tried to locally rename the test class, but it was somehow copied instead of moved. Sorry for this mess :) ).

mposolda commented:
@keycloak-ci-bot test

pedroigor commented:
@keycloak-ci-bot test

mposolda commented:
@keycloak-ci-bot test
